### PR TITLE
add being able to pass a `CacheFlags()` to `isprecompiled` to verify against a set of custom flags

### PR DESCRIPTION
**Pull request by KristofferC**
This makes it easier to also update `isprecompiled` to check for staleness with custom flags

This is the second take of https://github.com/JuliaLang/julia/pull/53316.

### CODE REVIEW

File: base/loading.jl [hunk 1]
```diff
@@ -1449,6 +1449,135 @@ end
...
+    debug_level = Int((f >> 1) & 3)
+    check_bounds = Int((f >> 3) & 3)
+    inline = Bool((f >> 5) & 1)
+    opt_level = Int((f >> 6) & 3) # define OPT_LEVEL in statiddata_utils
+    CacheFlags(use_pkgimages, debug_level, check_bounds, inline, opt_level)
+end
+CacheFlags(f::Int) = CacheFlags(UInt8(f))
+CacheFlags() = CacheFlags(ccall(:jl_cache_flags, UInt8, ()))
+
+function _to_uint8(cf::CacheFlags)::UInt8
```
Comment (KristofferC) [inline base/loading.jl:1472]: This is new

File: base/loading.jl [hunk 2]
```diff
@@ -1449,6 +1449,135 @@ end
...
+    f = UInt8(0)
+    f |= cf.use_pkgimages << 0
+    f |= cf.debug_level << 1
+    f |= cf.check_bounds << 3
+    f |= cf.inline << 5
+    f |= cf.opt_level << 6
+    return f
+end
+
+function _to_cmd(flags::Base.CacheFlags)
```
Comment (KristofferC) [inline base/loading.jl:1482]: This is new

**Review by vchuravy — APPROVED** [review]
I like this approach! Makes a lot of sense.

The only thing that is missing is `-C` handling which is not part of the CacheFlags, but impacts the cache in a similar way (if not more) than `-O`

It's the ImageTarget, but that is a more verbose representation than what we normally pass in.

**Comment by KristofferC** [comment]
> The only thing that is missing is -C handling which is not part of the CacheFlags, but impacts the cache in a similar way (if not more) than -O

 But are you saying this PR is ok as is or does something need to be added?

**Comment by IanButterworth** [comment]
Sounds good to me, but haven't done a code review.

**Comment by KristofferC** [comment]
I had to scale this back a bit (see latest commit) because we cannot recreate the command line flags from the `CacheFlags` object (`--pkgimages=existing`) so it became a bit awkward. Now you can verify a compile file against a `CacheFlags` at least.

**Comment by KristofferC** [comment]
Let's try this then.

File: src/jl_exported_funcs.inc [hunk 3]
```diff
@@ -387,6 +387,7 @@
     XX(jl_raise_debugger) \
     XX(jl_readuntil) \
     XX(jl_cache_flags) \
+    XX(jl_match_cache_flags_current) \
```
Comment (vtjnash) [inline src/jl_exported_funcs.inc:390]: Minor nit: this file is the list of functions in julia.h that are intended to be the eventual stable API for use by C programs. You don't need to add functions here merely to use them from Julia, and doing so can add confusion to what the C API is.

**Pull request merged**